### PR TITLE
Add QuantityInput control that supports formatting and parsing quantity values supported by QuantityFormatter.

### PR DESCRIPTION
**Pull request by bsteinbk**
The main focus of this PR is to provide a React Input Control for formatted quantity values.  The QuantityInput control can support any of the QuantityTypes supported by the QuantityFormatter. The QuantityFormatter defines the "persistence" unit and the format. For length, the "persistence" unit is meter. The Format specification defines how the value is converted to a string.  If the imperial length format is set and the value 1m is entered the displayed text will be `3'-3 3/8"` while the onChange value be 1.0 (since persistence unit is meter).

### CODE REVIEW

**Review by DanEastBentley — APPROVED** [review]

**Review by calebmshafer — APPROVED** [review]
Approving for the changelogs.

@bsteinbk the Docs build is going to fail at the moment. There was an issue in the pipeline itself, I should have that fixed in a couple minutes and will re-queue the build.

File: ui/components/src/test/inputs/ParsedInput.test.tsx [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    await MockRender.App.startup();
+  });
+
+  after(async () => {
+    await MockRender.App.shutdown();
+    TestUtils.terminateUiComponents();
+    Object.defineProperty(IModelApp, "requestNextAnimation", rnaDescriptorToRestore);
+  });
+
+  afterEach(cleanup);
```
Comment (GerardasB) [inline ui/components/src/test/inputs/ParsedInput.test.tsx:65]: Not necessary as cleanup is already invoked in 
https://github.com/imodeljs/imodeljs/blob/c0b588d4b2ae8d9600dc10a3fe33ba9d326e1681/ui/scripts/setup-tests.js#L103-L104

File: ui/components/src/test/inputs/QuantityInput.test.tsx [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    await MockRender.App.startup();
+  });
+
+  after(async () => {
+    await MockRender.App.shutdown();
+    TestUtils.terminateUiComponents();
+    Object.defineProperty(IModelApp, "requestNextAnimation", rnaDescriptorToRestore);
+  });
+
+  afterEach(cleanup);
```
Comment (GerardasB) [inline ui/components/src/test/inputs/QuantityInput.test.tsx:33]: Not necessary as cleanup is already done in 
https://github.com/imodeljs/imodeljs/blob/c0b588d4b2ae8d9600dc10a3fe33ba9d326e1681/ui/scripts/setup-tests.js#L103-L104

File: ui/components/src/test/inputs/ParsedInput.test.tsx [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    const initialTemperature = 20;  // 20 C
+    const spyOnChange = sinon.spy();
+
+    const wrapper = render(<ParsedInput onChange={spyOnChange} initialValue={initialTemperature} formatValue={formatCelsiusValue} parseString={parseStringToCelsius} />);
+    expect(wrapper).not.to.be.undefined;
+    const input = (await wrapper.findByTestId("components-parsed-input")) as HTMLInputElement;
+    expect(input.value).to.eq("20.0C");
+    fireEvent.change(input, { target: { value: "32F" } });
+    fireEvent.keyDown(input, { key: SpecialKey.Enter });
+    await TestUtils.flushAsyncOperations();
```
Comment (GerardasB) [inline ui/components/src/test/inputs/ParsedInput.test.tsx:77]: This is probably not necessary as there is no async code in ParsedInput.

File: ui/components/src/test/inputs/ParsedInput.test.tsx [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+  afterEach(cleanup);
+
+  it("should process format and parse function", async () => {
+    const initialTemperature = 20;  // 20 C
+    const spyOnChange = sinon.spy();
+
+    const wrapper = render(<ParsedInput onChange={spyOnChange} initialValue={initialTemperature} formatValue={formatCelsiusValue} parseString={parseStringToCelsius} />);
+    expect(wrapper).not.to.be.undefined;
+    const input = (await wrapper.findByTestId("components-parsed-input")) as HTMLInputElement;
```
Comment (GerardasB) [inline ui/components/src/test/inputs/ParsedInput.test.tsx:73]: This is rendered immediately, consider using `getBy` query.
```suggestion
    const input = wrapper.getByTestId("components-parsed-input") as HTMLInputElement;
```

File: ui/components/src/test/inputs/ParsedInput.test.tsx [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+  after(async () => {
+    await MockRender.App.shutdown();
+    TestUtils.terminateUiComponents();
+    Object.defineProperty(IModelApp, "requestNextAnimation", rnaDescriptorToRestore);
+  });
+
+  afterEach(cleanup);
+
+  it("should process format and parse function", async () => {
```
Comment (GerardasB) [inline ui/components/src/test/inputs/ParsedInput.test.tsx:67]: Test is no longer async after other comments are resolved.
```suggestion
```

File: ui/components/src/test/inputs/QuantityInput.test.tsx [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    const spyOnChange = sinon.spy();
+    const wrapper = render(<QuantityInput initialValue={initialLength} quantityType={QuantityType.Length} onQuantityChange={spyOnChange} />);
+    expect(wrapper).not.to.be.undefined;
+    const input = await wrapper.findByTestId("components-parsed-input");
+    fireEvent.change(input, { target: { value: "2.5" } });
+    expect(spyOnChange).not.to.have.been.called;
+    fireEvent.keyDown(input, { key: SpecialKey.Enter });
+    await TestUtils.flushAsyncOperations();
+    expect(spyOnChange).to.have.been.called;
+  });
```
Comment (GerardasB) [inline ui/components/src/test/inputs/QuantityInput.test.tsx:46]: Make test synchronous by removing `flushAsyncOperations` and replacing `findBy` with `getBy` queries.

File: ui/components/src/test/inputs/QuantityInput.test.tsx [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    const initialValue = input.value;
+    fireEvent.change(input, { target: { value: "2.5" } });
+    fireEvent.keyDown(input, { key: SpecialKey.Escape });
+    fireEvent.keyDown(input, { key: SpecialKey.Enter });
+    await TestUtils.flushAsyncOperations();
+    expect(spyOnChange).not.to.have.been.called;  // value did not change after ESC was pressed
+    const currentValue = input.value;
+    expect(initialValue).to.eq(currentValue);
+
+    IModelApp.quantityFormatter.onActiveUnitSystemChanged.emit({ useImperial: true });
```
Comment (GerardasB) [inline ui/components/src/test/inputs/QuantityInput.test.tsx:65]: Why is this needed after all assertions of the test are done?

File: ui/components/src/test/inputs/QuantityInput.test.tsx [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+  });
```
Comment (GerardasB) [inline ui/components/src/test/inputs/QuantityInput.test.tsx:66]: Make test synchronous by removing `flushAsyncOperations` and replacing `findBy` with `getBy` queries.

File: ui/components/src/test/inputs/QuantityInput.test.tsx [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    await TestUtils.flushAsyncOperations();
+    expect(input.classList.contains("components-parsed-input-has-error")).to.be.true;
+    fireEvent.keyDown(input, { key: SpecialKey.Escape });
+    expect(spyOnChange).not.to.have.been.called;  // value did not change after ESC was pressed
+    const currentValue = input.value;
+    await TestUtils.flushAsyncOperations();
+    expect(input.classList.contains("components-parsed-input-has-error")).to.be.false;
+    expect(initialValue).to.eq(currentValue);
+
+  });
```
Comment (GerardasB) [inline ui/components/src/test/inputs/QuantityInput.test.tsx:88]: Make test synchronous by removing `flushAsyncOperations` and replacing `findBy` with `getBy` queries.

File: ui/components/src/ui-components/inputs/ParsedInput.scss [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+$quantity-inputs-focus-alert-gradient: $quantity-focus-boxshadow-alert-gradient1 0px 1px 1px, $quantity-focus-boxshadow-alert-gradient2 0px 0px 0px 2px;
+
+.components-parsed-input {
+  width: auto;
+  &.components-parsed-input-has-error {
+    color: $buic-foreground-alert;
+    font-weight: bold;
+  }
+  &.components-parsed-input[disabled] {
+    color: var(--buic-foreground-body);
```
Comment (GerardasB) [inline ui/components/src/ui-components/inputs/ParsedInput.scss:18]: Consider using existing Sass variable.
```suggestion
    color: $buic-foreground-body;
```

File: ui/components/src/ui-components/inputs/ParsedInput.scss [hunk 11]
```diff
@@ -0,0 +1,22 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+@import "~@bentley/ui-core/lib/ui-core/style/themecolors";
+
+$quantity-focus-boxshadow-alert-gradient1:   rgba(var(--buic-accessory-alert-tint), 0);
+$quantity-focus-boxshadow-alert-gradient2:     rgba(var(--buic-accessory-alert), 0.65);
```
Comment (GerardasB) [inline ui/components/src/ui-components/inputs/ParsedInput.scss:8]:
```suggestion
$quantity-focus-boxshadow-alert-gradient1:   rgba($buic-accessory-alert-tint, 0);
$quantity-focus-boxshadow-alert-gradient2:     rgba($buic-accessory-alert, 0.65);
```

File: ui/components/src/ui-components/inputs/ParsedInput.tsx [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+  readonly?: boolean;
+}
+
+/** Generic Input component that requires formatting and parsing functions to be passed in as props.
+ * @beta
+ */
+export const ParsedInput = React.forwardRef<HTMLInputElement, ParsedInputProps>(
+  function ParsedInput({ initialValue, formatValue, parseString, readonly, className, style, onChange }, ref) {
+    const currentValueRef = React.useRef(initialValue);
+    const isMountedRef = React.useRef(false);
```
Comment (GerardasB) [inline ui/components/src/ui-components/inputs/ParsedInput.tsx:37]: `isMountedRef` flag is probably not needed in this component, because `updateQuantityValueFromString` is only called in `input` events?
```suggestion
```

File: ui/components/src/ui-components/inputs/ParsedInput.tsx [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+      } else {
+        setHasBadInput(true);
+      }
+    }, [formatValue, onChange, parseString]);
+
+    const handleOnBlur = React.useCallback((event: React.FocusEvent<HTMLInputElement>) => {
+      updateQuantityValueFromString(event.target.value);
+    }, [updateQuantityValueFromString]);
+
+    function onInputKeyDown(event: React.KeyboardEvent<HTMLInputElement>): void {
```
Comment (GerardasB) [inline ui/components/src/ui-components/inputs/ParsedInput.tsx:94]: Unify event handler names as currently they are all different: `handleInputChange` (_Input_ w/o _On_), `handleOnBlur` (_On_ w/o _Input_) and `onInputKeyDown` (_Input_ w/o _On_, but with _on_ prefix).
```suggestion
    function handleKeyDown(event: React.KeyboardEvent<HTMLInputElement>) {
```

File: ui/components/src/ui-components/inputs/ParsedInput.tsx [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+              setHasBadInput(false);
+            }
+          }
+        }
+      } else {
+        setHasBadInput(true);
+      }
+    }, [formatValue, onChange, parseString]);
+
+    const handleOnBlur = React.useCallback((event: React.FocusEvent<HTMLInputElement>) => {
```
Comment (GerardasB) [inline ui/components/src/ui-components/inputs/ParsedInput.tsx:90]: Unify handler names.
```suggestion
    const handleBlur = React.useCallback((event: React.FocusEvent<HTMLInputElement>) => {
```

File: ui/components/src/ui-components/inputs/ParsedInput.tsx [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+      // istanbul ignore else
+      if (initialValue !== currentValueRef.current) {
+        currentValueRef.current = initialValue;
+        lastFormattedValueRef.current = formatValue(initialValue);
+        setFormattedValue(lastFormattedValueRef.current);
+        setHasBadInput(false);
+      }
+    }, [formatValue, initialValue])
+
+    const handleInputChange = React.useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
```
Comment (GerardasB) [inline ui/components/src/ui-components/inputs/ParsedInput.tsx:60]: Unify handler names.
```suggestion
    const handleChange = React.useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
```

Comment (GerardasB) [inline ui/components/src/ui-components/inputs/ParsedInput.tsx:60]: Unify inconsistent handler memoization: `handleInputChange` and `handleOnBlur` are memoized while `onInputKeyDown` is not.
In this case there is no performance benefit to memoize (other than keeping codebase consistent) because `Input` component is not a PureComponent.

File: ui/components/src/ui-components/inputs/ParsedInput.tsx [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+      isMountedRef.current = true;
+      return () => {
+        isMountedRef.current = false;
+      }
+    }, [])
+
+    // See if new initialValue props have changed since component mounted
+    React.useEffect(() => {
+      // istanbul ignore else
+      if (initialValue !== currentValueRef.current) {
```
Comment (GerardasB) [inline ui/components/src/ui-components/inputs/ParsedInput.tsx:52]: Formatted value will not be updated if if new formatter - `formatValue` is passed in.

File: ui/components/src/ui-components/inputs/ParsedInput.tsx [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+        return;
+
+      const parseResults = (parseString(strVal));
+      // istanbul ignore else
+      if (!parseResults.parseError) {
+        // istanbul ignore else
+        if (undefined !== parseResults.value) {
+          // istanbul ignore else
+          if (currentValueRef.current !== parseResults.value) {
+            currentValueRef.current = parseResults.value as number;
```
Comment (GerardasB) [inline ui/components/src/ui-components/inputs/ParsedInput.tsx:75]: Something other than number could have been parsed, probably need a type guard.

File: ui/components/src/ui-components/inputs/QuantityInput.tsx [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+          return { parseError: `${UiComponents.translate("QuantityInput.NoParserDefined")}${parseResult.status}` };
+        }
+      }
+      // istanbul ignore next
+      return { parseError: UiComponents.translate("QuantityInput.NoParserDefined") };
+    };
+
+    const handleQuantityChange = React.useCallback((newValue: number) => {
+      onQuantityChange && onQuantityChange(newValue);
+    }, [onQuantityChange]);
```
Comment (GerardasB) [inline ui/components/src/ui-components/inputs/QuantityInput.tsx:66]: Just pass `onQuantityChange` down to `ParsedInput` (onQuantityChange prop is also always defined).
```suggestion
```

File: ui/components/src/ui-components/inputs/QuantityInput.tsx [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+
+/** Input control that allows users to input a quantity and show the formatted string that represents the value.
+ * @beta
+ */
+export const QuantityInput = React.forwardRef<HTMLInputElement, QuantityProps>(
+  function QuantityInput({ initialValue, quantityType, readonly, className, style, onQuantityChange }, ref) {
+    const [formatterSpec, setFormatterSpec] = React.useState(() => IModelApp.quantityFormatter.findFormatterSpecByQuantityType(quantityType));
+    const [parserSpec, setParserSpec] = React.useState(() => IModelApp.quantityFormatter.findParserSpecByQuantityType(quantityType));
+
+    const formatValue = React.useCallback((value: number) => {
```
Comment (GerardasB) [inline ui/components/src/ui-components/inputs/QuantityInput.tsx:40]: If `ParsedInput` is adjusted to parse when new formatter is passed in it might make sense to memoize this.

Comment (bsteinbk) [reply ui/components/src/ui-components/inputs/QuantityInput.tsx:40]: Will look at this in the future if needed. I am having issues using useMemo on component that uses React.forwardRef.

File: ui/components/src/test/inputs/QuantityInput.test.tsx (hunk 7 above)
Comment (bsteinbk) [reply ui/components/src/test/inputs/QuantityInput.test.tsx]: Updated to use IModelApp.quantityFormatter.useImperialFormats to format both metric and imperial values and to test the results.

File: ui/components/src/ui-components/inputs/ParsedInput.scss (hunk 10 above)
Comment (bsteinbk) [reply ui/components/src/ui-components/inputs/ParsedInput.scss]: Remove unused code.

File: ui/components/src/ui-components/inputs/ParsedInput.tsx (hunk 12 above)
Comment (bsteinbk) [reply ui/components/src/ui-components/inputs/ParsedInput.tsx:37]: Keeping in for now will do further investigation when type editor is constructed to use this component.

File: ui/components/src/ui-components/inputs/ParsedInput.tsx (hunk 16 above)
Comment (bsteinbk) [reply ui/components/src/ui-components/inputs/ParsedInput.tsx]: Will handle updated formatValue function now.

**Review by GerardasB — APPROVED** [review]

**Review by ramanujam-raman — APPROVED** [review]
Reviewed the QuantityFormatter changes.
